nixos/modules/virtualisation/railcar.nix

```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.railcar;
+in
+{
+  options.services.railcar = {
+    enable = mkEnableOption "railcar";
+
+    containers = mkOption {
+      default = {};
+      type = with types; loaOf (submodule ({ name, config, ... }: {
+        options = {
+          cmd = mkOption {
+            type = types.package;
+            description = "Command to run inside the container";
+          };
+
+          packages = mkOption {
+            type = with types; loaOf (package);
+            default = [];
+            description = "A set of additional packages to provide";
+          };
+
+          mounts = mkOption {
+            type = with types; loaOf (path);
+            default = [];
+            description = "A set of mounted directories inside the container";
+          };
+
+          runType = mkOption {
+            default = "oneshot";
+            description = "The systemd service run type";
+          };
+
+          os = mkOption {
+            default = "linux";
+            description = "OS type of the container";
+          };
+
+          arch = mkOption {
+            default = "x86_64";
+            description = "Computer architecture type of the container";
+          };
+        };   
+      }));
+    };
+
+    stateDir = mkOption {
+      default = ''/var/railcar'';
+      description = "Railcar persistent state directory";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.railcar;
+      description = "Railcar package to use";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services = flip mapAttrs' cfg.containers (container: containerConfig:
+      nameValuePair "${container}" {
+        enable = true;
+        serviceConfig = let
```

Comment:
I'm not sure I can? The only `let` in that block that is globally applicable is the `ociRunner`, which, as you pointed out, should probably be inlined.

Or am I misunderstanding how this would work?

Reply:
I was thinking something like this:

```nix
let
  cfg = config.services.railcar;

  containers = mapAttrs pkgs.ociTools.buildContainer cfg.containers;
```